Realm/ObjectStore/impl/async_query.cpp

```diff
@@ -0,0 +1,265 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2015 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "async_query.hpp"
+
+#include "realm_coordinator.hpp"
+#include "results.hpp"
+
+using namespace realm;
+using namespace realm::_impl;
+
+AsyncQuery::AsyncQuery(Results& target)
+: m_target_results(&target)
+, m_realm(target.get_realm().shared_from_this())
+, m_sort(target.get_sort())
+, m_sg_version(Realm::Internal::get_shared_group(*m_realm).get_version_of_current_transaction())
+{
+    Query q = target.get_query();
+    m_query_handover = Realm::Internal::get_shared_group(*m_realm).export_for_handover(q, MutableSourcePayload::Move);
+}
+
+AsyncQuery::~AsyncQuery()
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
```

Comment:
If unregister() is called while the query is running then the AsyncQuery will be destructed on the worker thread, so it needs a memory barrier to ensure the writes done in unregister() are visible. The actual lock is unnecessary, but thread sanitizer currently doesn't support explicit memory barriers.